File: backend/routers/subscription.py
import os
import requests
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from typing import Optional
from database import execute_query

router = APIRouter()

LOUVIN_API_KEY = os.getenv("LOUVIN_API_KEY", "lv_e837b09fd975494bb25ab6eaeffef179")
LOUVIN_BASE_URL = "https://api.louvin.dev"

class SubscriptionRequest(BaseModel):
    uid: str
    customer_email: str
    customer_name: str

@router.post("/create")
def create_subscription(req: SubscriptionRequest):
    try:
        # 1. Hit Louvin API to create transaction for 60.000 IDR
        payload = {
            "amount": 60000,
            "payment_type": "qris",
            "customer_name": req.customer_name,
            "customer_email": req.customer_email,
            "description": "Upgrade Akun Pro - Pasar UMKM",
            "reference": f"PRO-{req.uid}"
        }
        
        headers = {
            "Content-Type": "application/json",
            "x-api-key": LOUVIN_API_KEY
        }
        
        response = requests.post(f"{LOUVIN_BASE_URL}/create-transaction", json=payload, headers=headers)
        data = response.json()
        
        if not data.get("success"):
            raise Exception(data.get("error", "Gagal menghubungi Louvin API"))
            
        return {"status": "success", "data": data}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.get("/check-status")
def check_subscription_status(transaction_id: str, uid: str):
    try:
        # Check status with Louvin
        headers = {"x-api-key": LOUVIN_API_KEY}
        response = requests.get(f"{LOUVIN_BASE_URL}/check-status?id={transaction_id}", headers=headers)
        data = response.json()
        
        if not data.get("success"):
            raise Exception(data.get("error", "Gagal cek status transaksi"))
            
        status = data.get("transaction", {}).get("status")
        
        if status == "settled":
            # Update user in database to PRO and add 10 credits limit
            query = """
                UPDATE users 
                SET isSubscribed = 1, aiLimit = aiLimit + 10, aiSearchLimit = aiSearchLimit + 10, aiDescLimit = aiDescLimit + 10, aiAnalysisLimit = aiAnalysisLimit + 10
                WHERE uid = ?
            """
            execute_query(query, (uid,))
            return {"status": "success", "payment_status": "settled", "message": "Akun berhasil diupgrade ke PRO"}
            
        return {"status": "success", "payment_status": status}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

File: backend_go/routers/subscription.go
package routers

import (
	"backend/database"
	"backend/models"
	"bytes"
	"encoding/json"
	"io"
	"net/http"
	"os"

	"github.com/gofiber/fiber/v2"
)

var (
	LouvinBaseURL = "https://api.louvin.dev"
)

func getLouvinAPIKey() string {
	key := os.Getenv("LOUVIN_API_KEY")
	if key == "" {
		return "lv_e837b09fd975494bb25ab6eaeffef179"
	}
	return key
}

func CreateSubscription(c *fiber.Ctx) error {
	var req models.SubscriptionRequest
	if err := c.BodyParser(&req); err != nil {
		return c.Status(400).JSON(fiber.Map{"detail": err.Error()})
	}

	payload := map[string]interface{}{
		"amount":         60000,
		"payment_type":   "qris",
		"customer_name":  req.CustomerName,
		"customer_email": req.CustomerEmail,
		"description":    "Upgrade Akun Pro - Pasar UMKM",
		"reference":      "PRO-" + req.UID,
	}

	payloadBytes, _ := json.Marshal(payload)

	httpReq, err := http.NewRequest("POST", LouvinBaseURL+"/create-transaction", bytes.NewBuffer(payloadBytes))
	if err != nil {
		return c.Status(500).JSON(fiber.Map{"detail": err.Error()})
	}

	httpReq.Header.Set("Content-Type", "application/json")
	httpReq.Header.Set("x-api-key", getLouvinAPIKey())

	client := &http.Client{}
	resp, err := client.Do(httpReq)
	if err != nil {
		return c.Status(500).JSON(fiber.Map{"detail": err.Error()})
	}
	defer resp.Body.Close()

	body, _ := io.ReadAll(resp.Body)

	var data map[string]interface{}
	if err := json.Unmarshal(body, &data); err != nil {
		return c.Status(500).JSON(fiber.Map{"detail": "Failed to parse Louvin response"})
	}

	if success, ok := data["success"].(bool); !ok || !success {
		errMsg := "Gagal menghubungi Louvin API"
		if msg, ok := data["error"].(string); ok {
			errMsg = msg
		}
		return c.Status(500).JSON(fiber.Map{"detail": errMsg})
	}

	return c.JSON(fiber.Map{
		"status": "success",
		"data":   data,
	})
}

func CheckSubscriptionStatus(c *fiber.Ctx) error {
	transactionID := c.Query("transaction_id")
	uid := c.Query("uid")

	httpReq, err := http.NewRequest("GET", LouvinBaseURL+"/check-status?id="+transactionID, nil)
	if err != nil {
		return c.Status(500).JSON(fiber.Map{"detail": err.Error()})
	}
	httpReq.Header.Set("x-api-key", getLouvinAPIKey())

	client := &http.Client{}
	resp, err := client.Do(httpReq)
	if err != nil {
		return c.Status(500).JSON(fiber.Map{"detail": err.Error()})
	}
	defer resp.Body.Close()

	body, _ := io.ReadAll(resp.Body)

	var data map[string]interface{}
	if err := json.Unmarshal(body, &data); err != nil {
		return c.Status(500).JSON(fiber.Map{"detail": "Failed to parse Louvin response"})
	}

	if success, ok := data["success"].(bool); !ok || !success {
		errMsg := "Gagal cek status transaksi"
		if msg, ok := data["error"].(string); ok {
			errMsg = msg
		}
		return c.Status(500).JSON(fiber.Map{"detail": errMsg})
	}

	txData, _ := data["transaction"].(map[string]interface{})
	status, _ := txData["status"].(string)

	if status == "settled" {
		query := `
			UPDATE users 
			SET isSubscribed = 1, aiLimit = aiLimit + 10, aiSearchLimit = aiSearchLimit + 10, aiDescLimit = aiDescLimit + 10, aiAnalysisLimit = aiAnalysisLimit + 10
			WHERE uid = ?
		`
		_, err := database.DB.Exec(query, uid)
		if err != nil {
			return c.Status(500).JSON(fiber.Map{"detail": err.Error()})
		}
		return c.JSON(fiber.Map{
			"status":         "success",
			"payment_status": "settled",
			"message":        "Akun berhasil diupgrade ke PRO",
		})
	}

	return c.JSON(fiber.Map{
		"status":         "success",
		"payment_status": status,
	})
}

func SetupSubscriptionRoutes(router fiber.Router) {
	router.Post("/create", CreateSubscription)
	router.Get("/check-status", CheckSubscriptionStatus)
}
